### PR TITLE
feat(config-con): go config loader

### PR DESCRIPTION
**Pull request by alfredosa**

### CODE REVIEW

File: configloader/configloader.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+	if err := l.loadFields(v, t, l.prefix); err != nil {
+		return fmt.Errorf("failed to load configuration: %w", err)
+	}
+	return nil
+}
+
+// loadFields recursively processes struct fields and loads values from environment variables.
+func (l *Loader) loadFields(v reflect.Value, t reflect.Type, prefix string) error {
+	if !v.IsValid() {
+		return fmt.Errorf("struct should be an initialized pointer")
```
Comment (hadrienk) [inline configloader/configloader.go:81]: Remove? 

What happens if: 
```go
type Foo struct {
  Field *Bar
} 
```

Comment (hadrienk) [reply configloader/configloader.go:81]: See what Unmarshal does?
https://pkg.go.dev/encoding#TextUnmarshaler

File: configloader/configloader_test.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+				t.Errorf("Loader() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+
+			if err != nil {
+				return
+			}
+
+			// Compare results
+			if got.Environment != tt.want.Environment {
```
Comment (hadrienk) [inline configloader/configloader_test.go:107]: Use `assert.Equal`?

File: configloader/configloader_test.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+			if got.Timeout != tt.want.Timeout {
+				t.Errorf("Timeout = %v, want %v", got.Timeout, tt.want.Timeout)
+			}
+		})
+	}
+}
+
+func TestFails_OnConfigNotBeingStruct(t *testing.T) {
+	conf := AppConfig{}
+	assert.Error(t, Load(conf), "config should fail as it's not a pointer")
```
Comment (hadrienk) [inline configloader/configloader_test.go:146]: Add nil pointer test
```go
var conf *AppConfig
```

File: configloader/configloader_test.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+type AppConfig struct {
+	Environment string
+	Debug       bool
+	Database    DatabaseConfig
+	AllowedIps  []string
+	MaxRetries  int
+	Timeout     float64
+}
+
+func TestLoad(t *testing.T) {
```
Comment (hadrienk) [inline configloader/configloader_test.go:27]: Check how Unmarshal deals with custom type (https://pkg.go.dev/encoding/json#Unmarshaler)

Comment (hadrienk) [reply configloader/configloader_test.go:27]: Implement special types as well: 
URL,
Time
Duration
etc.

File: configloader/configloader.go (hunk 1 above)
Comment (hadrienk) [reply configloader/configloader.go:81]:
```go
type Loader struct {
	prefix, tag string

	// Type of conversion the struct field will take. Default is SnakeCase.
	fieldConversion func(string) string

	// Custom GetEnv function.
	env func(string) string

	handler []func(value reflect.Value, envValue string) error
}

var SkipHandlerErr = errors.New("unsupported type")
func DurationHandler(value reflect.Type, envValue string) (any,error) {
	if value.Kind() != time.Duration {
		return nil, SkipHandlerErr
	}
	// If type != time.Duration return error or false.
	duration, err := time.ParseDuration(envValue)
	if err != nil {
		return nil, err
	}
	return duration, nil
}
```

File: configloader/configloader.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+func defaultLoader() *Loader {
+	return &Loader{
+		fieldConversion: strcase.ToScreamingSnake,
+		env:             os.Getenv,
+	}
+}
+
+type Option func(*Loader)
+
+// WithPrefix sets the prefix for environment variable names.
```
[the remark this answers is not in the document]
Comment (alfredosa) [reply configloader/configloader.go:32]: thank you :)

File: configloader/configloader.go (hunk 1 above)
Comment (alfredosa) [reply configloader/configloader.go:95]: Please see new impl @hadrienk

File: configloader/configloader_test.go (hunk 3 above)
Comment (alfredosa) [reply configloader/configloader_test.go:107]: Added

File: configloader/configloader.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+	}
+}
+
+// WithEnv sets a custom environment variable lookup function.
+func WithEnv(env func(string) string) Option {
+	return func(l *Loader) {
+		l.env = env
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (alfredosa) [reply configloader/configloader.go:65]: I'm still thinking how to do this. The lib will return a zero value struct by default. Would this maybe the scope of the user?

**Review by LinMAD — COMMENTED** [review]
Nice, don't have code related comments.

However would be nice to add better `godoc` to make the library user-friendly.
Especially describe how the struct tags work add some examples, especially if there are additional options maybe like (env:"X_NAME,optional").

File: configloader/configloader.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (geethu-mohan) [reply configloader/configloader.go:65]: hmm, I think that should be okay.

**Review by geethu-mohan — DISMISSED** [review]

**Comment by hadrienk** [comment]
> However would be nice to add better `godoc` to make the library user-friendly.
> Especially describe how the struct tags work add some examples, especially if there are additional options maybe like (env:"X_NAME,optional").

I agree, @alfredosa let's merge this PR and create a task for documentation so it doesn't drag too much. I recommend using LLM for this. Passing both the code and tests it should generate pretty good documentation in no time, and leave time for us to improve it.

**Comment by alfredosa** [comment]
> > However would be nice to add better `godoc` to make the library user-friendly.
> > Especially describe how the struct tags work add some examples, especially if there are additional options maybe like (env:"X_NAME,optional").
> 
> I agree, @alfredosa let's merge this PR and create a task for documentation so it doesn't drag too much. I recommend using LLM for this. Passing both the code and tests it should generate pretty good documentation in no time, and leave time for us to improve it.

@hadrienk  @LinMAD ticket for all go docs created

**Pull request merged**